src/Symfony/Component/Security/Http/Authentication/AuthenticatorManager.php

```diff
@@ -0,0 +1,257 @@
+            // as its support is relying on the (initialized) token in the TokenStorage.
+            if (false === $authenticator->supports($request)) {
+                if (null !== $this->logger) {
+                    $this->logger->debug('Skipping the "{authenticator}" authenticator as it did not support the request.', ['authenticator' => \get_class($authenticator)]);
+                }
+                continue;
+            }
+
+            $response = $this->executeAuthenticator($key, $authenticator, $request);
+            if (null !== $response) {
+                if (null !== $this->logger) {
+                    $this->logger->debug('The "{authenticator}" authenticator set the response. Any later authenticator will not be called', ['authenticator' => \get_class($authenticator)]);
+                }
+
+                return $response;
+            }
+        }
+
+        return null;
+    }
+
+    private function executeAuthenticator(string $uniqueAuthenticatorKey, AuthenticatorInterface $authenticator, Request $request): ?Response
+    {
+        try {
+            if (null !== $this->logger) {
+                $this->logger->debug('Calling getCredentials() on authenticator.', ['firewall_key' => $this->providerKey, 'authenticator' => \get_class($authenticator)]);
+            }
+
+            // allow the authenticator to fetch authentication info from the request
+            $credentials = $authenticator->getCredentials($request);
+
+            if (null === $credentials) {
+                throw new \UnexpectedValueException(sprintf('The return value of "%1$s::getCredentials()" must not be null. Return false from "%1$s::supports()" instead.', \get_class($authenticator)));
+            }
+
+            // authenticate the credentials (e.g. check password)
+            $token = $this->authenticateViaAuthenticator($authenticator, $credentials);
+
+            if (null !== $this->logger) {
+                $this->logger->info('Authenticator successful!', ['token' => $token, 'authenticator' => \get_class($authenticator)]);
+            }
+
+            // success! (sets the token on the token storage, etc)
+            $response = $this->handleAuthenticationSuccess($token, $request, $authenticator);
+            if ($response instanceof Response) {
+                return $response;
+            }
+
+            if (null !== $this->logger) {
+                $this->logger->debug('Authenticator set no success response: request continues.', ['authenticator' => \get_class($authenticator)]);
+            }
+
+            return null;
+        } catch (AuthenticationException $e) {
+            // oh no! Authentication failed!
+            $response = $this->handleAuthenticationFailure($e, $request, $authenticator);
+            if ($response instanceof Response) {
+                return $response;
+            }
+
+            return null;
```

Comment:
There are 2 log messages on the "success" side (`Authenticator successful` and `Authenticator set no success response`) that could be repeated down here for failure.

Reply:
See `handleAuthenticationFailure()` below :)